src/callbacks_step/amr.jl

```diff
@@ -282,14 +282,92 @@ function (amr_callback::AMRCallback)(u_ode::AbstractVector, mesh::TreeMesh,
     coarsened_original_cells = Int[]
   end
 
- # Return true if there were any cells coarsened or refined, otherwise false
- has_changed = !isempty(refined_original_cells) || !isempty(coarsened_original_cells)
- if has_changed # TODO: Taal decide, where shall we set this?
-  # don't set it to has_changed since there can be changes from earlier calls
-  mesh.unsaved_changes = true
- end
-
- return has_changed
+  # Return true if there were any cells coarsened or refined, otherwise false
+  has_changed = !isempty(refined_original_cells) || !isempty(coarsened_original_cells)
+  if has_changed # TODO: Taal decide, where shall we set this?
+    # don't set it to has_changed since there can be changes from earlier calls
+    mesh.unsaved_changes = true
+  end
+
+  if has_changed && mpi_isparallel()
+    old_mpi_ranks_per_cell = copy(mesh.tree.mpi_ranks)
+    # Load balancing
+    partition!(mesh)
+
+    # Retain current solution data
+    old_n_elements = nelements(dg, cache)
+    old_cell_ids = copy(cache.elements.cell_ids)
+    old_u_ode = copy(u_ode)
+    GC.@preserve old_u_ode begin # OBS! If we don't GC.@preserve old_u_ode, it might be GC'ed
+      old_u = wrap_array(old_u_ode, mesh, equations, dg, cache)
+
+      # Get new list of leaf cells
+      leaf_cell_ids = local_leaf_cells(mesh.tree)
+
+      # re-initialize elements container
+      @unpack elements = cache
+      resize!(elements, length(leaf_cell_ids))
+      init_elements!(elements, leaf_cell_ids, mesh, dg.basis.nodes)
+
+      resize!(u_ode, nvariables(equations) * nnodes(dg)^ndims(mesh) * nelements(dg, cache))
+      u = wrap_array(u_ode, mesh, equations, dg, cache)
+
+      # Find elements that will change their rank and send their data to the new rank
+      for old_element_id in 1:old_n_elements
+        cell_id = old_cell_ids[old_element_id]
+        if cell_id ∉ leaf_cell_ids
+          # Send element data to new rank, use cell_id as tag (non-blocking)
+          dest = mesh.tree.mpi_ranks[cell_id]
+          MPI.Isend(vec(old_u[:, .., old_element_id]), dest, cell_id, mpi_comm())
+        end
+      end
+
+      # Loop over all elements in new container and either copy them from old container 
+      # or receive them with MPI
+      for element_id in 1:nelements(dg, cache)
+        cell_id = elements.cell_ids[element_id]
+        if cell_id in old_cell_ids
+          old_element_id = searchsortedfirst(old_cell_ids, cell_id)
+          # Copy old element data to new element container
+          @views u[:, .., element_id] .= old_u[:, .., old_element_id]
+        else
+          # Receive old element data
+          src = old_mpi_ranks_per_cell[cell_id]
+          MPI.Recv!(@view(u[:, .., element_id]), src, cell_id, mpi_comm())
```

Comment:
I think this blocking `Recv!` can potentially cause serialization issues. That is, if the process that is sending needs to send a lot of data to the preceding rank and only little date to the following rank. On the following rank, the `Recv` will have to wait until it's corresponding `Isend` is called, thus not being able to process the remaining elements in the meantime.

If you can think of an not too convoluted implementation that fixes this (e.g., by using `MPI.Irecv`s somehow), maybe we should try to change it. OTOH, this may be a case of premature optimization, so please hold off if it is too complicated (in the sense that it would take you more than a few lines and half an hour of coding).

Reply:
I initially received the data non-blocking, stored it in a buffer and then copied it to `u`. As I needed to put a `MPI_Wait` immediately after this line, I just changed it to blocking and didn't think about it anymore.

Now that I don't use any buffers, wouldn't it be sufficient to just make this `Recv!` non-blocking? The receiving rank would go on with the loop copying the following old elements while the part of `u` for this one element remains unchanged. Once the rank receives the data, the `MPI.Recv!` will write it into `u` for the missing element.

Reply:
Yes, this sounds like it could work! And it's a minor change, so go for it!